ql/src/java/org/apache/hadoop/hive/ql/exec/util/Retryable.java

```diff
@@ -0,0 +1,212 @@
+        } else {
+          return callable.call();
+        }
+      } catch (Exception e) {
+        if (this.failOn.stream().noneMatch(k -> e.getClass().equals(k))
+          && this.retryOn.stream().anyMatch(k -> e.getClass().isAssignableFrom(k))) {
+          if (elapsedTimeInSeconds(startTime) + delay > this.totalDurationInSeconds) {
+            // case where waiting would go beyond max duration. So throw exception and return
+            throw e;
+          }
+          sleep(delay);
+          //retry case. compute next sleep time
+          delay = getNextDelay(delay, currentCapturedException, e);
+          // reset current captured exception.
+          currentCapturedException = e;
+        } else {
+          // Exception cannot be retried on. Throw exception and return
+          throw e;
+        }
+      }
+    }
+  }
+
+  private void sleep(long seconds) {
+    try {
+      Thread.sleep(seconds * 1000);
+    } catch (InterruptedException e) {
+      // no-op.. just proceed
+    }
+  }
+
+  private long getNextDelay(long currentDelay, final Exception previousException, final Exception currentException) {
+    if (previousException != null && !previousException.getClass().equals(currentException.getClass())) {
+      // New exception encountered. Returning initial delay for next retry.
+      return this.initialDelayInSeconds;
+    }
+
+    if (currentDelay <= 0) { // in case initial delay was set to 0.
+      currentDelay = MINIMUM_DELAY_IN_SEC;
+    }
+
+    currentDelay *= this.backOff;
+    if (this.maxJitterInSeconds > 0) {
+      currentDelay += new Random().nextInt(this.maxJitterInSeconds);
+    }
+
+    if (currentDelay > this.maxRetryDelayInSeconds) {
+      currentDelay = this.maxRetryDelayInSeconds;
+    }
+
+    return  currentDelay;
+  }
+
+  private long elapsedTimeInSeconds(long fromTimeMillis) {
+    return (System.currentTimeMillis() - fromTimeMillis)/ 1000;
+  }
+
+  public static class Builder {
+    private final Retryable runnable = new Retryable();
+    public Builder() {
+    }
+
+    public Builder withHiveConf(HiveConf conf) {
+      runnable.totalDurationInSeconds = conf.getTimeVar(HiveConf.ConfVars.REPL_RETRY_TOTAL_DURATION, TimeUnit.SECONDS);
+      runnable.initialDelayInSeconds = conf.getTimeVar(HiveConf.ConfVars.REPL_RETRY_INTIAL_DELAY, TimeUnit.SECONDS);
+      runnable.maxRetryDelayInSeconds = conf.getTimeVar(HiveConf.ConfVars
+        .REPL_RETRY_MAX_DELAY_BETWEEN_RETRIES, TimeUnit.SECONDS);
+      runnable.backOff = conf.getFloatVar(HiveConf.ConfVars.REPL_RETRY_BACKOFF_COEFFICIENT);
+      runnable.maxJitterInSeconds = (int) conf.getTimeVar(HiveConf.ConfVars.REPL_RETRY_JITTER, TimeUnit.SECONDS);
+      return this;
+    }
+
+    public Retryable build() {
+      return runnable;
+    }
+
+    public Builder withTotalDuration(long maxDuration) {
+      runnable.totalDurationInSeconds = maxDuration;
+      return this;
+    }
+
+    // making this thread safe as it appends to list
```

Comment:
Do we really  need it synchronized as the build process is single thread operation, no?

Reply:
If we use a common retryable it can be used in a multi threaded way as well. Right now we are not using it but keeping it synchronized will keep it thread safe and allow case where you create retryable once and pass the common retry params and then call excute from different places.